### ISSUE TITLE
code

### ISSUE THREAD

**Issue by hm37-blip** [hm37-blip/Miscanthus-Phenotype-Clustering#2]
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from google.colab import drive

drive.mount('/content/drive')

file_path = '/content/drive/My Drive/salinity/Data_for_all_4_replicates.xlsx'
df = pd.read_excel(file_path, na_values=['NA', 'NA ', '.', ' '])

features = [
    'SaltInjury_Post', 'SaltInjury_Recovery',
    'LeafSenescence_Post', 'LeafSenescence_Recovery',
    'LeafNo_Post', 'LeafNo_Recovery',
    'SPAD_Post', 'SPAD_Recovery'
]

df['Entry'] = df['Entry'].astype(str).str.strip()

for col in features:
    df[col] = pd.to_numeric(df[col], errors='coerce')

df_grouped = df.groupby('Entry')[features].mean()
df_clean = df_grouped.dropna()

scaler = StandardScaler()
X_scaled = scaler.fit_transform(df_clean)
kmeans = KMeans(n_clusters=4, random_state=42, n_init=10)
df_clean['Cluster'] = kmeans.fit_predict(X_scaled)

pca = PCA(n_components=2)
principal_components = pca.fit_transform(X_scaled)
df_clean['PC1'] = principal_components[:, 0]
df_clean['PC2'] = principal_components[:, 1]

plt.figure(figsize=(12, 8))

sns.scatterplot(
    x='PC1', y='PC2',
    hue='Cluster',
    data=df_clean,
    palette='viridis',
    s=60, alpha=0.6, edgecolor='w'
)

targets = {
    'UI10-00107': {'search': 'Illinois', 'label': 'UI10-00107 (Control)'},
    'UI11-00031': {'search': 'PF30022',  'label': 'PF30022 (Sensitive)'},
    'PMS-009':    {'search': 'PMS-009',  'label': 'PMS-009 (Tolerant)'}
}

for key, info in targets.items():
    # Search for the entry name that contains the specific substring
    matches = [name for name in df_clean.index if info['search'] in name]

    if matches:
        # Pick the first match found
        entry_name = matches[0]
        point = df_clean.loc[entry_name]

        plt.scatter(point['PC1'], point['PC2'], s=300, color='red', marker='*', edgecolors='black', zorder=10)
        plt.text(point['PC1']+0.2, point['PC2']+0.2, info['label'], fontsize=11, fontweight='bold', color='black',
                 bbox=dict(facecolor='white', alpha=0.8, edgecolor='none', boxstyle='round,pad=0.3'))
    else:
        print(f"Warning: Could not find any entry matching '{info['search']}'")

plt.title('K-Means Clustering of Salt Stress Traits (PCA Reduced)\nk=4 Clusters', fontsize=15)
plt.xlabel(f'Principal Component 1 ({pca.explained_variance_ratio_[0]*100:.1f}%)')
plt.ylabel(f'Principal Component 2 ({pca.explained_variance_ratio_[1]*100:.1f}%)')
plt.legend(title='Cluster', bbox_to_anchor=(1.05, 1), loc='upper left')
plt.grid(True, linestyle='--', alpha=0.5)
plt.tight_layout()
plt.show()

print("\n--- Cluster Details (Mean Values of Traits) ---")
print(df_clean.groupby('Cluster')[features].mean())

print("\n--- Number of Entries per Cluster ---")
print(df_clean['Cluster'].value_counts())
